### DISCUSSION TITLE
how to config AI for selfhost

### DISCUSSION THREAD

**Discussion by IamTaoChen** [toeverything/AFFiNE#6996] [General]
Can we use AI in the self-host instance?

**Comment by wanghkkk** [comment]
I have the same question

**Comment by vuxng** [comment]
Turn on the copilot plugin at 
`.affine/self-host/config/affine.js`

**Reply by lennon101** [reply]
> Turn on the copilot plugin at
> `.affine/self-host/config/affine.js`

nice find and thanks for this! I've gone ahead and uncommented the copilot plugin and added my OpenAI apikey (see below). And I in the Affine logs I can see the CoPilot is registered. But I'm still getting an error. Do I need to get a FAL and an unsplashKey as well? 

affine.js
```
// /* Copilot Plugin */
AFFiNE.use('copilot', {
  openai: {
    apiKey: 'sk-proj-<withheld>',
  },
  // fal: {
  //   apiKey: 'your-key',
  // },
  // unsplashKey: 'your-key',
  // storage: {
  //   provider: 'cloudflare-r2',
  //   bucket: 'copilot',
  // }
})
```

Affine logs: 
```
[Nest] 117  - 06/26/2024, 2:23:10 AM VERBOSE [CacheInterceptor] cache blobs:6ebf69e5-33b9-4e3b-9809-a99de7acab75 miss
[Nest] 117  - 06/26/2024, 2:23:18 AM   DEBUG [CopilotProviderService] Copilot provider openai registered, capabilities: text-to-text, text-to-embedding, text-to-image, image-to-text
```

UI error when trying to use AI:
![image](https://github.com/toeverything/AFFiNE/assets/1709944/ac138e1d-08cd-407f-8e79-bf9a933ccddf)

**Reply by Bottleface** [reply]
@lennon101 were you ever able to get this working?

**Comment by AndreasWidmer** [comment]
For everybody interested: this works with the self hosted version. You have to use (a paid) Open AI account with a **User API keys (Legacy)** - not the new project API keys! 

In the affine.js config set:

```
AFFiNE.use('copilot', {
   openai: {
           apiKey: 'your-key',
   },
//   fal: {
//     apiKey: 'your-key',
//   },
//   unsplashKey: 'your-key',
//   storage: {
//     provider: 'cloudflare-r2',
//     bucket: 'copilot',
//   }
 })
```

**Reply by campbellmcgregor** [reply]
any idea how to remove the 10 usage limit?

**Reply by madratzz** [reply]
You can modify it in the postgres database

**Reply by pyrox82** [reply]
Can I also choose which model is used? I think by default it's gpt-4o. But I rather want to use gpt-4o-mini.

**Reply by JoeBuildsStuff** [reply]
You might need to modify the model selection in the postgres table 

<img width="959" alt="image" src="https://github.com/user-attachments/assets/3e4bf8fe-0c9b-4bf2-a16f-0afcdc929583">

**Reply by richardy2012** [reply]
At this stage, modifications from the database are no longer effective, and restrictions are made directly at the code level. Only a series of built-in large models such as gpt-4o and gpt-4o-mini are supported, so don't bother with it.

**Reply by Let1fer** [reply]
Is there a way to edit the models used? My provider provides access under slightly different names

**Comment by Nishan-Vivek** [comment]
Looks like openai don't allow new accounts to generate legacy API keys anymore. Anyway to get this to work with the project api keys?

**Comment by hnykda** [comment]
I made it work with the latest 0.20 version: https://github.com/toeverything/AFFiNE/discussions/10549 . My openai API key starts with `sk-svcacct--`

**Comment by pierdevara** [comment]
i found another workaround, i use LiteLLM as a llm proxy, converts over 200 LLMs in openai compatible api calls. In the admin UI you have the option to give a fake name to the LLM, as gpt-4o, so it will be recognize by AFFine
```
AFFiNE.use('copilot', {
  openai: {
    baseURL: 'http://localhost:4000',
    apiKey: 'sk-xxxxxxxxxxxxxxx',
  }
});
```

**Reply by Steven-Low** [reply]
> i found another workaround, i use LiteLLM as a llm proxy, converts over 200 LLMs in openai compatible api calls. In the admin UI you have the option to give a fake name to the LLM, as gpt-4o, so it will be recognize by AFFine `AFFiNE.use('copilot', { openai: { baseURL: 'http://localhost:4000', apiKey: 'sk-xxxxxxxxxxxxxxx', } });`

Really appreciate the help, that solved it! 💘
however the UI require setting up postgresql so instead I just declare the litellm_config.yaml:
```
model_list:
  - model_name: gpt-4o
    litellm_params:
      model: openrouter/anthropic/claude-3.5-haiku
      api_key: os.environ/OPENROUTER_API_KEY
```

**Comment by floli** [comment]
Does that approach still works with 0.21? I have set:

```
AFFiNE.use('copilot', {
  openai: {
    apiKey: process.env.COPILOT_OPENAI_API_KEY,
  },
})
```
and set the API key in the `.env` file. It used to work on 0.20, however, on 0.21 all AI functionality is gone.

Any idea?

**Reply by tareeko** [reply]
I'd love to know if they're planning to add it back.

**Reply by sudesh0sudesh** [reply]
I think this new config.json appeared after 21 is the answer for new configs
https://github.com/toeverything/AFFiNE/blob/canary/.docker/selfhost/config.example.json

**Reply by floli** [reply]
That's right, my `config.json` now looks like

```js
{
  "$schema": "https://github.com/toeverything/affine/releases/latest/download/config.schema.json",
  "server": {
    "name": "domain.de",
    "host": "affine.domain.de",
    "https": true
  },
  "copilot": {
    "enabled": true,
    "providers.openai": {
      "apiKey": "sk-proj-xxx"
    }
  }
}
```

As already stated by someone else, while the AI in the side bar does not work, the AI on selected texts works just fine.

**Reply by sudesh0sudesh** [reply]
Admin section says they are working on AI

**Reply by alch3mistdev** [reply]
what are they working on though? UI just needs a config entry...? This sniffs of trying to force a subscription fee..

**Comment by laelhalawani** [comment]
This should be a feature.
Model selection and settings to paste in the API key in the UI.

**Reply by Harsh-Upadhayay** [reply]
I'm interested in implementing it, can anyone point me in the right direction for this contribution?

**Comment by hnykda** [comment]
Anyone giving it a shot on 0.23?

**Comment by mxvsh** [comment]
any update on this?

**Comment by HomingHamster** [comment]
What is FAL?

**Comment by pihomeserver** [comment]
Please read #11722 ([AI for selfhost](https://github.com/toeverything/AFFiNE/discussions/11722#top))
